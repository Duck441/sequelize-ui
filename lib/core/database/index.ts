import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: DatabaseCaseStyle
  nounForm: DatabaseNounForm
}

export enum DatabaseCaseStyle {
  Snake = 'SNAKE',
  Camel = 'CAMEL',
}

export enum DatabaseNounForm {
  Singular = 'SINGULAR',
  Plural = 'PLURAL',
}

export enum SqlDialect {
  MariaDb = 'MARIADB',
  MsSql = 'MSSQL',
  MySql = 'MYSQL',
  Postgres = 'POSTGRES',
  Sqlite = 'SQLITE',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'mariadb'
    case SqlDialect.MsSql:
      return 'mssql'
    case SqlDialect.MySql:
      return 'mysql'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return 'sqlite'
  }
}

export const varSqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MARIADB'
    case SqlDialect.MsSql:
      return 'MSSQL'
    case SqlDialect.MySql:
      return 'MYSQL'
    case SqlDialect.Postgres:
      return 'POSTGRES'
    case SqlDialect.Sqlite:
      return 'SQLITE'
  }
}

export const parseSqlDialect = (dialect: string): SqlDialect | null => {
  switch (dialect) {
    case 'mariadb':
      return SqlDialect.MariaDb
    case 'mssql':
      return SqlDialect.MsSql
    case 'mysql':
      return SqlDialect.MySql
    case 'postgres':
      return SqlDialect.Postgres
    case 'sqlite':
      return SqlDialect.Sqlite
    default:
      return null
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectPassword = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'Password1'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectHost = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'localhost'
    case SqlDialect.MsSql:
      return 'localhost'
    case SqlDialect.MySql:
      return 'localhost'
    case SqlDialect.Postgres:
      return 'localhost'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectStorage = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return null
    case SqlDialect.MsSql:
      return null
    case SqlDialect.MySql:
      return null
    case SqlDialect.Postgres:
      return null
    case SqlDialect.Sqlite:
      return '.tmp/data.db'
  }
}

export const sqlCurrentTimestamp = (): string => 'CURRENT_TIMESTAMP'
export const sqlCurrentDate = (): string => 'CURRENT_DATE'
export const sqlCurrentTime = (): string => 'CURRENT_TIME'

export const displayDatabaseNounForm = (nounForm: DatabaseNounForm): string => {
  switch (nounForm) {
    case DatabaseNounForm.Singular:
      return 'singular'
    case DatabaseNounForm.Plural:
      return 'plural'
  }
}

export const displayDatabaseCaseStyle = (caseStyle: DatabaseCaseStyle): string => {
  switch (caseStyle) {
    case DatabaseCaseStyle.Snake:
      return 'snake'
    case DatabaseCaseStyle.Camel:
      return 'camel'
  }
}

export const MAX_IDENTIFIER_LENGTH = 63
