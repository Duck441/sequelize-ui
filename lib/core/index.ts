export * from './codegen'
export * from './database'
export * from './files'
export * from './framework'
