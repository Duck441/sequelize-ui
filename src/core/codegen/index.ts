export function indent(depth: number, value: string, prefix = ''): string {
  return value
    .split('\n')
    .map((x) => prefix + ' '.repeat(depth) + x)
    .join('\n')
}

export function blank(): string {
  return ''
}

export type LinesOptions = {
  separator?: string
  depth?: number
  prefix?: string
}

export type Line = NestedArray<string | null>

const defaultLinesOptions: LinesOptions = {
  separator: '',
  depth: 0,
}
export function lines(
  xs: Line,
  { separator = '', depth = 0, prefix = '' }: LinesOptions = defaultLinesOptions,
): string {
  return indent(
    depth,
    flatten<string | null>(xs)
      .filter((x, i, arr): x is string => x !== null && !consecutiveBlank(arr, i))
      .join(`${separator}\n`),
    prefix,
  )
}

function consecutiveBlank(arr: Array<string | unknown>, i: number): boolean {
  const prev = arr[i - 1]
  const curr = arr[i]

  return (
    typeof curr === 'string' && typeof prev === 'string' && curr.trim() === '' && prev.trim() === ''
  )
}

type NestedArray<T> = Array<NestedArray<T> | T>
function flatten<T>(xs: NestedArray<T>): T[] {
  const output: T[] = []
  for (const x of xs) {
    if (Array.isArray(x)) {
      output.push(...flatten(x))
    } else {
      output.push(x)
    }
  }

  return output
}
