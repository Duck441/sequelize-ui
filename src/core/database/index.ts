import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: DatabaseCaseStyle
  nounForm: DatabaseNounForm
}

export enum DatabaseCaseStyle {
  Snake = 'snake',
  Camel = 'camel',
}

export enum DatabaseNounForm {
  Singular = 'singular',
  Plural = 'plural',
}

export enum SqlDialect {
  MariaDb = 'mariadb',
  MsSql = 'mssql',
  MySql = 'mysql',
  Postgres = 'postgres',
  Sqlite = 'sqlite',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MariaDB'
    case SqlDialect.MsSql:
      return 'SQL Server'
    case SqlDialect.MySql:
      return 'MySQL'
    case SqlDialect.Postgres:
      return 'PostgreSQL'
    case SqlDialect.Sqlite:
      return 'SQLite'
  }
}

export const sqlDialectEnvVar = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MARIADB'
    case SqlDialect.MsSql:
      return 'MSSQL'
    case SqlDialect.MySql:
      return 'MYSQL'
    case SqlDialect.Postgres:
      return 'POSTGRES'
    case SqlDialect.Sqlite:
      return 'SQLITE'
  }
}

export const parseSqlDialect = (dialect: string): SqlDialect | null => {
  switch (dialect) {
    case 'mariadb':
      return SqlDialect.MariaDb
    case 'mssql':
      return SqlDialect.MsSql
    case 'mysql':
      return SqlDialect.MySql
    case 'postgres':
      return SqlDialect.Postgres
    case 'sqlite':
      return SqlDialect.Sqlite
    default:
      return null
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectPassword = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'Password1'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectHost = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'localhost'
    case SqlDialect.MsSql:
      return 'localhost'
    case SqlDialect.MySql:
      return 'localhost'
    case SqlDialect.Postgres:
      return 'localhost'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectStorage = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return null
    case SqlDialect.MsSql:
      return null
    case SqlDialect.MySql:
      return null
    case SqlDialect.Postgres:
      return null
    case SqlDialect.Sqlite:
      return '.tmp/data.db'
  }
}

export const displayDatabaseNounForm = (nounForm: DatabaseNounForm): string => {
  switch (nounForm) {
    case DatabaseNounForm.Singular:
      return 'singular'
    case DatabaseNounForm.Plural:
      return 'plural'
  }
}

export const displayDatabaseCaseStyle = (caseStyle: DatabaseCaseStyle): string => {
  switch (caseStyle) {
    case DatabaseCaseStyle.Snake:
      return 'snake'
    case DatabaseCaseStyle.Camel:
      return 'camel'
  }
}

export const defaultDbOptions: DatabaseOptions = {
  sqlDialect: SqlDialect.Postgres,
  timestamps: true,
  caseStyle: DatabaseCaseStyle.Snake,
  nounForm: DatabaseNounForm.Plural,
}

export const MAX_IDENTIFIER_LENGTH = 63
