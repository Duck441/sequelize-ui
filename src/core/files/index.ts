export * from './fileSystem'
export * from './fileTree'
