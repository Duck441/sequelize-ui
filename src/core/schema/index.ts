export * from './association'
export * from './dataType'
export * from './schema'
export * from './validation'
