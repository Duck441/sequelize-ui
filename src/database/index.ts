import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: 'snake' | 'camel'
  nounForm: 'singular' | 'plural'
}

export enum SqlDialect {
  MariaDb = 'MARIADB',
  MsSql = 'MSSQL',
  MySql = 'MYSQL',
  Postgres = 'POSTGRES',
  Sqlite = 'SQLITE',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'mariadb'
    case SqlDialect.MsSql:
      return 'mssql'
    case SqlDialect.MySql:
      return 'mysql'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return 'sqlite'
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}
