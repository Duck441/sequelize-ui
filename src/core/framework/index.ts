import { DbOptions } from '@src/core/database'
import { DirectoryItem } from '@src/core/files'
import { Model, Schema } from '@src/core/schema'

export type GenerateArgs = {
  schema: Schema
  dbOptions: DbOptions
}

export enum ProjectType {
  Npm = 'NPM',
}

export interface Framework {
  displayName(): string
  generate(args: GenerateArgs): DirectoryItem
  projectType(): ProjectType
  defaultFile?(root: DirectoryItem): string | undefined
  defaultModelFile?(model: Model, root: DirectoryItem): string | undefined
  modelFromPath?(path: string, schema: Schema): Model | undefined
}
