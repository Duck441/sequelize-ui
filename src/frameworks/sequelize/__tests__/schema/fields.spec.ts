import { defaultDbOptions } from '@src/core/database'
import fieldsSchema from '@src/data/schemas/fields'
import { SequelizeFramework } from '../..'

describe('Sequelize Framework', () => {
  describe('schema - fields', () => {
    it('generates correct code', () => {
      const code = SequelizeFramework.generate({
        schema: fieldsSchema,
        dbOptions: defaultDbOptions,
      })
      expect(code).toMatchSnapshot()
    })
  })
})
