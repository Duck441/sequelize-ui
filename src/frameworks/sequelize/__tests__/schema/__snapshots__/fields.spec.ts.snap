// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequelize Framework schema - fields generates correct code 1`] = `
Object {
  "files": Array [
    Object {
      "files": Array [
        Object {
          "content": "import type { Sequelize, Model } from 'sequelize'
import { Category } from './Category'
import type { CategoryAttributes, CategoryCreationAttributes } from './Category'

export {
  Category
}

export type {
  CategoryAttributes,
  CategoryCreationAttributes
}

export function initModels(sequelize: Sequelize) {
  Category.initModel(sequelize)

  

  return {
    Category
  }
}
",
          "name": "index.ts",
          "type": "FILE",
        },
        Object {
          "content": "import Sequelize, { DataTypes, Model, Optional } from 'sequelize'

export interface CategoryAttributes {
  field?: string
  pkField?: string
  requiredField: string
  optionalField?: string
}

export type CategoryPk = 'pkField'
export type CategoryId = CategoryAttributes[CategoryPk]
export type CategoryCreationAttributes = Optional<CategoryAttributes, CategoryPk>

export class Category extends Model<CategoryAttributes, CategoryCreationAttributes> implements CategoryAttributes {
  public field?: string
  public readonly pkField?: string
  public requiredField!: string
  public optionalField?: string
  
  static initModel(sequelize: Sequelize.Sequelize): typeof Category {
    Category.init({
      field: {
        type: DataTypes.STRING
      },
      pkField: {
        type: DataTypes.STRING,
        primaryKey: true
      },
      requiredField: {
        type: DataTypes.STRING,
        allowNull: false
      },
      optionalField: {
        type: DataTypes.STRING,
        unique: true
      }
    }, {
      sequelize
    })
    
    return Category
  }
}
",
          "name": "Category.ts",
          "type": "FILE",
        },
      ],
      "name": "models",
      "type": "DIRECTORY",
    },
    Object {
      "content": "node_modules/
dist
npm-debug.log*
yarn-debug.log*
yarn-error.log*
.npm
.tmp
",
      "name": ".gitignore",
      "type": "FILE",
    },
    Object {
      "content": "import { Sequelize } from 'sequelize'

const db: Sequelize = new Sequelize({
  dialect: 'postgres',
  database: process.env.POSTGRES_DB_NAME || 'fields',
  username: process.env.POSTGRES_DB_USERNAME || 'postgres',
  password: process.env.POSTGRES_DB_PASSWORD || 'postgres',
  host: process.env.POSTGRES_DB_HOST || 'localhost',
  port: parseInt(process.env.POSTGRES_DB_PORT || '5432'),
  define: {
    underscored: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  }
})

export default db
",
      "name": "db.ts",
      "type": "FILE",
    },
    Object {
      "content": "{
  \\"name\\": \\"fields\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"\\",
  \\"main\\": \\"server.ts\\",
  \\"scripts\\": {
    \\"test\\": \\"echo \\\\\\"Error: no test specified\\\\\\" && exit 1\\",
    \\"build\\": \\"tsc\\",
    \\"start\\": \\"node ./dist/server.js\\"
  },
  \\"author\\": \\"\\",
  \\"license\\": \\"ISC\\",
  \\"dependencies\\": {
    \\"pg\\": \\"^8.5.1\\",
    \\"pg-hstore\\": \\"^2.3.3\\",
    \\"sequelize\\": \\"^6.3.5\\"
  },
  \\"devDependencies\\": {
    \\"@types/node\\": \\"^14.14.20\\",
    \\"@types/validator\\": \\"^13.1.3\\",
    \\"typescript\\": \\"^4.1.3\\"
  }
}

",
      "name": "package.json",
      "type": "FILE",
    },
    Object {
      "content": "import db from './db'
import { initModels } from './models'

initModels(db)

db.sync({ force: true })
",
      "name": "server.ts",
      "type": "FILE",
    },
    Object {
      "content": "{
    \\"compilerOptions\\": {
      \\"target\\": \\"ES2020\\",
      \\"module\\": \\"commonjs\\",
      \\"lib\\": [\\"ES2020\\"],
      \\"outDir\\": \\"dist\\",
      \\"strict\\": true,
      \\"moduleResolution\\": \\"node\\",
      \\"esModuleInterop\\": true,
      \\"forceConsistentCasingInFileNames\\": true
    }
  }
",
      "name": "tsconfig.json",
      "type": "FILE",
    },
  ],
  "name": "fields",
  "type": "DIRECTORY",
}
`;
